### PR TITLE
Discord Webhooks

### PR DESCRIPTION
**Pull request by dunakeyr**
Adding support for Discord webooks.

### CODE REVIEW

File: .env.template [hunk 1]
```diff
@@ -9,8 +11,8 @@
 # WALGREENS_X_XSRF_TOKEN=
 # WALGREENS_XSRF_TOKEN_COOKIE=
 
-CVS_ALLOW_LIST={"MO": ["SAINT LOUIS"], "KS": []}
-CVS_BLOCK_LIST={"MO": [], "KS": []}
+CVS_ALLOW_LIST={"MO": ["SAINT LOUIS"], "KS": [], "IA": []}
+CVS_BLOCK_LIST={"MO": [], "KS": [], "IA": []}
```
Comment (pcarn) [inline .env.template:15]: I'd like to leave additional states out of the template, people can add it to their own .env file.

File: README.md [hunk 2]
```diff
@@ -1,7 +1,7 @@
 # Vaccine Watch
 Bot to notify when vaccine appointments are available.
 
-Supports checking Hy-Vee, and sending notifications to Slack and [Twitter](https://twitter.com/kcvaccinewatch).
+Supports checking Hy-Vee, and sending notifications to Slack, Discord and [Twitter](https://twitter.com/kcvaccinewatch).
```
Comment (pcarn) [inline README.md:4]: Prefer oxford comma
```suggestion
Supports checking Hy-Vee, and sending notifications to Slack, Discord ,and [Twitter](https://twitter.com/kcvaccinewatch).
```

File: README.md [hunk 3]
```diff
@@ -63,7 +69,7 @@ Required Environment Variables:
   - CVS has a maximum of 25, vaccine-watch will use 25 for CVS if you set this higher.
 - `LATITUDE`: Latitude of the location to check (e.g. `39.1040382`)
 - `LONGITUDE`: Latitude of the location to check (e.g. `-94.5701803`)
-- `STATES`: JSON: Abbreviations of which states are in radius of your location (e.g. `["MO", "KS"]`)
+- `STATES`: JSON: Abbreviations of which states are in radius of your location (e.g. `["MO", "KS", "IA"]`)
```
Comment (pcarn) [inline README.md:72]: Same here, let's keep the example to just two states

File: notify/discord.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
+import json
+import logging
+import os
+import requests
+
+states = json.loads(os.environ["STATES"])
+
+
+def format_available_message(locations):
+    message = ":green_circle: Vaccine appointments available at {} location{}:".format(
```
Comment (pcarn) [inline notify/discord.py:10]: I don't use discord, but can you verify if this puts the emoji in the message correctly like slack does?

File: notify/discord.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+
+def send_message_to_discord(message):
+    webhook_url = os.environ["DISCORD_WEBHOOK_URL"]
+    data = {"content": message}
+    try:
+        result = requests.post(webhook_url, json=data)
+        result.raise_for_status()
+        logging.info("Payload delivered successfully, code {}.".format(result.status_code))
+    except requests.exceptions.HTTPError as err:
+        logging.exception(err)
```
Comment (pcarn) [inline notify/discord.py:70]: `logging.exception` will include the error details with it, you can remove `as err` from above and pass a message like
`logging.exception("Error sending message to discord")`

File: notify/discord.py [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+        "these" if len(locations) > 1 else "this",
+        "s" if len(locations) > 1 else "",
+    )
+    for location in locations:
+        if "earliest_appointment_day" in location:
+            if (
+                location["earliest_appointment_day"]
+                == location["latest_appointment_day"]
+            ):
+                day_string = " on *{}*".format(location["earliest_appointment_day"])
```
Comment (pcarn) [inline notify/discord.py:21]: Same with this formatting, does the markdown appear the same way as slack?
Example:

File: notify/discord.py (hunk 4 above)
Comment (dunakeyr) [reply notify/discord.py:11]: That it does. Below is example of posting.

image

Comment (dunakeyr) [reply notify/discord.py:11]: please note on the url %7Chere is fixed. This screenshot was before that change.

Comment (pcarn) [reply notify/discord.py:11]: `*test*` does bold in slack, it does italic in [Discord markdown](https://support.discord.com/hc/en-us/articles/210298617-Markdown-Text-101-Chat-Formatting-Bold-Italic-Underline-)
`**test**` does bold

Comment (pcarn) [reply notify/discord.py:11]: If possible, you could do a link like how slack has it:
https://stackoverflow.com/questions/63160401/how-to-create-hyperlink-in-discord-in-an-embed-in-general
image

If not possible, you're welcome to use the tinyurl shortener that twitter.py uses

Comment (dunakeyr) [reply notify/discord.py:11]: I did check, that markdown feature is not available in discord. I will put in a shortened url.

Comment (dunakeyr) [reply notify/discord.py:11]: updated...

image

File: .env.template (hunk 1 above)
Comment (dunakeyr) [reply .env.template]: change implemented.

File: README.md (hunk 2 above)
Comment (dunakeyr) [reply README.md]: change implemented.

File: README.md (hunk 3 above)
Comment (dunakeyr) [reply README.md]: change implemented.

File: notify/discord.py (hunk 5 above)
Comment (dunakeyr) [reply notify/discord.py]: change implemented, but double check as I may have misunderstood you.

File: notify/discord.py (hunk 4 above)
Comment (dunakeyr) [reply notify/discord.py:13]: changes implemented, please check formatting to make sure it looks good.

**Review by pcarn — APPROVED** [review]
Thanks!

**Pull request merged**